cpp/src/arrow/adapters/avro/adapter.cc

```diff
@@ -0,0 +1,325 @@
+        auto value_type = get_arrow_type(avro_schema_map_values(schema));
+        auto fields = {field("key", utf8()), field("value", value_type)};
+        return list(struct_(fields));
+      }
+
+      case AVRO_UNION: {
+        size = avro_schema_union_size(schema);
+        auto fields = std::vector<std::shared_ptr<Field>>();
+        std::vector<uint8_t> type_codes;
+
+        for (idx = 0; idx < size; idx++) {
+          child = avro_schema_union_branch(schema, idx);
+          auto f = field("_union_" + std::to_string(idx),
+                         get_arrow_type(child));
+          fields.push_back(f);
+          type_codes.push_back((uint8_t) idx);
+        };
+        return union_(fields, type_codes);
+      }
+
+      case AVRO_RECORD: {
+        size = avro_schema_record_size(schema);
+        auto fields = std::vector<std::shared_ptr<Field>>();
+        for (idx = 0; idx < size; idx++) {
+          child = avro_schema_record_field_get_by_index(schema, idx);
+          auto elem_type = get_arrow_type(child);
+          name = avro_schema_record_field_name(schema, idx);
+          auto f = field(name, elem_type);
+          fields.insert(fields.end(), f);
+        }
+        return struct_(fields);
+      }
+    }
+  }
+
+  Status generic_read(const avro_value_t val, ArrayBuilder *builder) {
+    // Generic avro type read dispatcher. Dispatches to the various specializations by AVRO_TYPE.
+    // This is used by the various readers for complex types"""
+    avro_type_t avro_type;
+    avro_type = avro_value_get_type(&val);
+
+    switch (avro_type) {
+      case AVRO_BOOLEAN:
+        return this->read_bool(val, builder);
+      case AVRO_INT32:
+        return this->read_int32(val, builder);
+      case AVRO_INT64:
+        return this->read_int64(val, builder);
+      case AVRO_FLOAT:
+        return this->read_float32(val, builder);
+      case AVRO_DOUBLE:
+        return this->read_float64(val, builder);
+      case AVRO_STRING:
+        return this->read_string(val, builder);
+      case AVRO_BYTES:
+        return this->read_binary(val, builder);
+      case AVRO_FIXED:
+        return this->read_fixed(val, builder);
+      case AVRO_MAP:
+        return this->read_map(val, builder);
+      case AVRO_RECORD:
+        return this->read_record(val, builder);
+      case AVRO_ARRAY:
+        return this->read_array(val, builder);
+      default:
+        std::stringstream ss;
+        ss << "Unhandled avro type: " << avro_type;
+        return Status::NotImplemented(ss.str());
+    }
+  }
+
+  Status read_string(const avro_value_t val, ArrayBuilder *builder) {
+    size_t strlen;
+    const char *c_str = NULL;
+    avro_value_get_string(&val, &c_str, &strlen);
+    return static_cast<StringBuilder *>(builder)->Append(c_str);
+  }
+
+  Status read_binary(const avro_value_t val, ArrayBuilder *builder) {
+    size_t strlen;
+    const char *c_str = NULL;
+    avro_value_get_bytes(&val, (const void **) &c_str, &strlen);
```

Comment:
Hm, if `avro_value_get_bytes`'s second argument is `void**` then you shouldn't have to cast, but if you definitely do need to then you should do this with `reinterpret_cast<void**>(&c_str)`.

This should also fail because you can't mutate anything of type `const void**` unless someone is doing something nasty like casting away the `const` with `const_cast`.